ding/model/template/ebm.py

```diff
@@ -0,0 +1,488 @@
+
+    def _sample(self, obs: torch.Tensor, num_samples: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Helper method for drawing action samples from the uniform random distribution
+        and tiling observations to the same shape as action samples.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        action_bounds = self.action_bounds.cpu().numpy()
+        size = (obs.shape[0], num_samples, action_bounds.shape[1])
+        action_samples = np.random.uniform(action_bounds[0, :], action_bounds[1, :], size=size)
+        action_samples = torch.as_tensor(action_samples, dtype=torch.float32).to(self.device)
+        tiled_obs = unsqueeze_repeat(obs, num_samples, 1)
+        return tiled_obs, action_samples
+
+    @staticmethod
+    @torch.no_grad()
+    def _get_best_action_sample(obs: torch.Tensor, action_samples: torch.Tensor, ebm: nn.Module):
+        """Return target with highest probability (lowest energy).
+        obs: (B, N, O), action_samples: (B, N, A)
+        return: (B, A).
+        """
+        # (B, N)
+        energies = ebm.forward(obs, action_samples)
+        probs = F.softmax(-1.0 * energies, dim=-1)
+        # (B, )
+        best_idxs = probs.argmax(dim=-1)
+        return action_samples[torch.arange(action_samples.size(0)), best_idxs]
+
+    def set_action_bounds(self, action_bounds: np.ndarray):
+        self.action_bounds = torch.as_tensor(action_bounds, dtype=torch.float32).to(self.device)
+
+    @abstractmethod
+    def sample(self, obs: torch.Tensor, ebm: nn.Module) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Create tiled observations and sample counter-negatives for feeding to the InfoNCE objective.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def infer(self, obs: torch.Tensor, ebm: nn.Module) -> torch.Tensor:
+        """Optimize for the best action conditioned on the current observation.
+        obs: (B, O)
+        return: (B, A).
+        """
+        raise NotImplementedError
+
+
+@STOCHASTIC_OPTIMIZER_REGISTRY.register('dfo')
+class DFO(StochasticOptimizer):
+
+    def __init__(
+        self,
+        noise_scale: float = 0.33,
+        noise_shrink: float = 0.5,
+        iters: int = 3,
+        train_samples: int = 8,
+        inference_samples: int = 16384,
+        device: str = 'cpu',
+    ):
+        self.action_bounds = None
+        self.noise_scale = noise_scale
+        self.noise_shrink = noise_shrink
+        self.iters = iters
+        self.train_samples = train_samples
+        self.inference_samples = inference_samples
+        self.device = device
+
+    def sample(self, obs: torch.Tensor, ebm: nn.Module) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Create tiled observations and sample counter-negatives for feeding to the InfoNCE objective.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        del ebm
```

Comment:
why del here

Reply:
`del` as a reminder that although `ebm` is an argument it is not used in DFO sampling. This is copied from the pytorch version of DFO implementation.